Layers/Renderers/Chunking/ChunkedLayerCompositionRenderer.gd

```diff
@@ -0,0 +1,143 @@
+@export var detailed_load_distance := 2000.0
+
+
+func custom_chunk_setup(chunk): pass
+
+
+func _ready():
+	super._ready()
+	
+	for x in range(-extent, extent + 1):
+		for y in range(-extent, extent + 1):
+			var chunk = chunk_scene.instantiate()
+
+			chunk.position = Vector3(x * chunk_size, 0.0, y * chunk_size)
+			chunk.size = chunk_size
+			
+			custom_chunk_setup(chunk)
+			
+			# Start as a downgraded (low resolution) chunk
+			chunk.downgrade()
+			
+			chunks.append(chunk)
+	
+	for chunk in chunks:
+		$Chunks.add_child(chunk)
+
+
+func is_new_loading_required(position_diff: Vector3) -> bool:
+	return not waiting_to_apply and (abs(position_diff.x) > chunk_size or abs(position_diff.z) > chunk_size)
+
+
+func full_load():
+	for chunk in chunks:
+		chunk.position_diff = Vector3.ZERO
+		
+		chunk.build(center[0], center[1])
+
+
+func adapt_load(_diff: Vector3):
+	super.adapt_load(_diff)
+	
+	# Because this function is called in a thread, the player position might change while the loop
+	#  below is running. In order to get consistent results, we need to cache one definitive
+	#  player position here.
+	var player_x = position_manager.center_node.position.x
+	var player_z = position_manager.center_node.position.z
+	
+	for chunk in chunks:
+		var changed = false
+
+		if chunk.position.x - player_x >= chunk_size * extent + chunk_size / 2.0:
+			chunk.position_diff.x = -chunk_size * extent * 2 - chunk_size
+			changed = true
+		elif chunk.position.x - player_x <= -chunk_size * extent - chunk_size / 2.0:
+			chunk.position_diff.x = chunk_size * extent * 2 + chunk_size
+			changed = true
+		
+		if chunk.position.z - player_z >= chunk_size * extent + chunk_size / 2.0:
+			chunk.position_diff.z = -chunk_size * extent * 2 - chunk_size
+			changed = true
+		elif chunk.position.z - player_z <= -chunk_size * extent - chunk_size / 2.0:
+			chunk.position_diff.z = chunk_size * extent * 2 + chunk_size
+			changed = true
+		
+		if changed:
+			chunk.downgrade()
+			chunk.build(center[0], center[1])
+	
+	waiting_to_apply = true
+	call_deferred("apply_new_data")
+
+
+func get_nearest_downgraded_chunk(query_position: Vector3, max_distance: float):
+	var nearest_distance = INF
+	var nearest_chunk
+	
+	for chunk in chunks:
+		if not chunk.is_upgraded:
+			var distance = Vector2(chunk.position.x, chunk.position.z).distance_to(Vector2(query_position.x, query_position.z))
+			
+			if distance < nearest_distance and distance < max_distance:
+				nearest_distance = distance
+				nearest_chunk = chunk
+	
+	return nearest_chunk
+
+
+func refine_load():
+	if waiting_to_apply: return
+	
+	super.refine_load()
+	
+	var any_change_done = false
+	
+	# Downgrade chunks which are now too far away
+	for chunk in chunks:
+		var distance = Vector2(chunk.position.x, chunk.position.z).distance_to(Vector2(position_manager.center_node.position.x, position_manager.center_node.position.z))
+		if chunk.is_upgraded and distance > detailed_load_distance:
+			chunk.downgrade()
```

Comment:
I think  a quick comment or maybe documentation of the class could be beneficial to understand that chunks are an LoD system that `down`- or `upgrade`s LoD levels.

Reply:
Also in my opinion up- and downgrade are somewhat misleading terms as a lower LoD usually means a higher resolution. I suggest a more specific function name like `downgrade_quality/resolution`. In any case, a clarification would be beneficial